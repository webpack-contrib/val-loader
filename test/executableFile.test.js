import path from "path";

import { getCompiler, compile, readAsset, normalizeErrors } from "./helpers";

describe("executableFile option", () => {
  it("should work with commonjs format", async () => {
    const compiler = getCompiler(
      "executableFileEntry.js",
      {},
      {
        module: {
          rules: [
            {
              test: /\.(json)$/i,
              rules: [
                {
                  loader: require.resolve("./helpers/helperLoader.js"),
                },
                {
                  loader: require.resolve("../src"),
                  options: {
                    executableFile: path.resolve(
                      __dirname,
                      "fixtures",
                      "executableFile.js",
                    ),
                  },
                },
              ],
            },
            {
              test: /\.json$/i,
              type: "asset/resource",
            },
          ],
        },
      },
    );
    const stats = await compile(compiler);

    expect(readAsset("val-loader.js", compiler, stats)).toMatchSnapshot(
      "result",
    );
    expect(normalizeErrors(stats.compilation.warnings)).toMatchSnapshot(
      "warnings",
    );
    expect(normalizeErrors(stats.compilation.errors)).toMatchSnapshot("errors");
  });

  // TODO jest have not good support for ES modules for testing it, tested manually
  it.skip("should work with ES modules format", async () => {
    const compiler = getCompiler(
      "executableFileEntry.js",
      {},
      {
        module: {
          rules: [
            {
              test: /\.(json)$/i,
              rules: [
                {
                  loader: require.resolve("./helpers/helperLoader.js"),
                },
                {
                  loader: require.resolve("../src"),
                  options: {
                    executableFile: path.resolve(
                      __dirname,
                      "fixtures",
                      "executableFileES.mjs",
                    ),
                  },
                },
              ],
            },
            {
              test: /\.json$/i,
              type: "asset/resource",
            },
          ],
        },
      },
    );
    const stats = await compile(compiler);

    expect(readAsset("val-loader.js", compiler, stats)).toMatchSnapshot(
      "result",
    );
    expect(normalizeErrors(stats.compilation.warnings)).toMatchSnapshot(
      "warnings",
    );
    expect(normalizeErrors(stats.compilation.errors)).toMatchSnapshot("errors");
  });

  it("should emit error", async () => {
    const compiler = getCompiler(
      "executableFileEntry.js",
      {},
      {
        module: {
          rules: [
            {
              test: /\.(json)$/i,
              rules: [
                {
                  loader: require.resolve("./helpers/helperLoader.js"),
                },
                {
                  loader: require.resolve("../src"),
                  options: {
                    executableFile: path.resolve(
                      __dirname,
                      "fixtures",
                      "error-require.js",
                    ),
                  },
                },
              ],
            },
            {
              test: /\.json$/i,
              type: "asset/resource",
            },
          ],
        },
      },
    );
    const stats = await compile(compiler);

    expect(normalizeErrors(stats.compilation.warnings)).toMatchSnapshot(
      "warnings",
    );
    expect(normalizeErrors(stats.compilation.errors)).toMatchSnapshot("errors");
  });
});
