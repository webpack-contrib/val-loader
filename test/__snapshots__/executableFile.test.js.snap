// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`executableFile option should emit error: errors 1`] = `
[
  "ModuleBuildError: Module build failed (from \`replaced original path\`):
Error: Unable to require "/test/fixtures/error-require.js": Error: This is a typical require() error",
]
`;

exports[`executableFile option should emit error: warnings 1`] = `[]`;

exports[`executableFile option should work with commonjs format: errors 1`] = `[]`;

exports[`executableFile option should work with commonjs format: result 1`] = `
"{
  "content": "module.exports = 315360000000",
  "map": null,
  "meta": null,
  "dependencies": [
    "test/fixtures/data.json"
  ],
  "contextDependencies": [],
  "buildDependencies": []
}"
`;

exports[`executableFile option should work with commonjs format: warnings 1`] = `[]`;
